File: src/lc_41_first_missing_pos.c
#include <stdio.h>
#include <stdlib.h>

void print_array(int* arr, int arr_size){

    int i = 0;
    for(i; i < arr_size; i++){
        printf("%d, ", arr[i]);
    }
    printf("\n");
}

int firstMissingPositive(int* nums, int numsSize){

    int i;
    int adj_i;
    int output = 1;

    //Set nums < 1 to 0.
    for( i = 0; i < numsSize; i++){
        if(nums[i] < 1){
            nums[i] = 0;
        }
    }

    //Mark sequential numbers found starting from 1.
    //Mark by making positive number a negative.
    for(i = 0; i < numsSize; i++){

        adj_i = abs(nums[i]) - 1;
        
        if( nums[i] == 0){
            continue;
        }
        else if( (adj_i < numsSize) && (nums[adj_i] == 0)){
            nums[adj_i] = abs(nums[i]) * -1;
        }
        else if (adj_i < numsSize) {
            //
            nums[adj_i] = abs(nums[adj_i]) * -1;
        }
    }

    //Find missing positive
    for(i = 0; i < numsSize; i++){
        
        if(nums[i] >= 0){
            return output;
        }
        output++;
    }

    return output;

}

File: src/lc_41_first_missing_pos.py
class Solution:

    def __init__(self):
        pass

    def firstMissingPositive(self, nums):

        is_1_seen = False

        #Set invalid numbers in array to 1
        for i in range(0,len(nums)):

            if(nums[i] == 1):
                is_1_seen = True

            elif ((nums[i] < 1) or (nums[i] > len(nums))):
                nums[i] = 1;

        if not is_1_seen:
            return 1

        #Use array as hash table and mark seen numbers as negative
        for i in range(0,len(nums)):
            if (abs(nums[i]) == len(nums)):
                nums[0] = abs(nums[i]) * -1
            else:
                nums[abs(nums[i])] = abs(nums[abs(nums[i])]) * -1

        #Search for missing positive
        for i in range(1,len(nums)):
            if (nums[i] > 0):
                return i

        #Search for missing positive
        if(nums[0] > 0):
            return len(nums)
        else:
            return len(nums) + 1
